wonk/models.py

```diff
@@ -2,35 +2,77 @@
 
 import copy
 import json
+import math
 import re
-from typing import Any, Dict, List, Set, Tuple, Union
-
-from .constants import ACTION_KEYS, JSON_ARGS, RESOURCE_KEYS, StatementKey
+from dataclasses import dataclass, field
+from hashlib import sha256
+from typing import Any, Dict, Generator, List, Set, Tuple, Union
+
+from .constants import (
+    ACTION_KEYS,
+    JSON_ARGS,
+    MAX_MANAGED_POLICY_SIZE,
+    RESOURCE_KEYS,
+    PolicyKey,
+    StatementKey,
+)
+from .exceptions import UnshrinkablePolicyError
 
 Statement = Dict[str, Any]
-Policy = Dict[str, Any]
 
 
+@dataclass(frozen=True)
```

Comment:
What was the motivation for making this frozen?

Reply:
I realized, during a painful debugging sessions ("this can't _possibly_ be happening!"), that I'd lazily let a couple of functions mutate statements in-place. This forces us to be more strict about explicitly returning updated copies, making it easier to follow data as it flows through. For the same reason, I made `Policy` objects immutable from the start to avoid re-learning that lesson.